File: upload_bridge.js
/**
 * AURA STREAM - CAMERA UPLOAD BRIDGE (NODE.JS)
 * 
 * Works out-of-the-box in Node.js 18+ (uses native fetch & FormData).
 * Watches a folder for new photos and uploads them to the server.
 */

const fs = require('fs');
const path = require('path');

// ==============================================================================
# CONFIGURATION
// ==============================================================================
// 1. Path to the folder where your camera uploads photos (via FTP, SD card, etc.)
const WATCH_FOLDER = './camera_photos';

// 2. Your API Upload URL (obtained from the dashboard integration panel)
//    Example: "http://localhost:3000/api/upload/your-event-id?secret=your-secret"
const UPLOAD_URL = 'http://localhost:3000/api/upload/replace-this-with-event-id?secret=your-secret';

// 3. File containing the log of already uploaded files (to prevent re-upload on restart)
const HISTORY_FILE = '.uploaded_history.log';
// ==============================================================================

// Ensure watch folder exists
if (!fs.existsSync(WATCH_FOLDER)) {
  fs.mkdirSync(WATCH_FOLDER, { recursive: true });
  console.log(`[+] Created watch folder: ${path.resolve(WATCH_FOLDER)}`);
} else {
  console.log(`[+] Watching folder: ${path.resolve(WATCH_FOLDER)}`);
}

// Load uploaded history
let uploadedFiles = new Set();
if (fs.existsSync(HISTORY_FILE)) {
  const content = fs.readFileSync(HISTORY_FILE, 'utf8');
  uploadedFiles = new Set(content.split('\n').map(line => line.trim()).filter(Boolean));
  console.log(`[*] Loaded ${uploadedFiles.size} previously uploaded files from history.`);
}

function saveToHistory(filePath) {
  fs.appendFileSync(HISTORY_FILE, filePath + '\n', 'utf8');
}

console.log(`[+] API Endpoint: ${UPLOAD_URL}`);
console.log(`[+] History stored in: ${path.resolve(HISTORY_FILE)}`);
console.log('[*] Waiting for new photos... (Press Ctrl+C to stop)');
console.log('------------------------------------------------------------');

// Polling folder every 1.5 seconds (more robust across platforms than fs.watch)
setInterval(async () => {
  try {
    const files = fs.readdirSync(WATCH_FOLDER);
    
    for (const filename of files) {
      if (filename.startsWith('.')) continue; // ignore hidden files
      
      const filePath = path.join(WATCH_FOLDER, filename);
      const absPath = path.resolve(filePath);
      
      // Check if it is a file and has a valid image extension
      const stats = fs.statSync(filePath);
      if (stats.isFile()) {
        const ext = filename.toLowerCase();
        const isValidImage = ['.jpg', '.jpeg', '.png', '.heic', '.heif', '.cr2', '.cr3', '.nef', '.arw'].some(suffix => ext.endsWith(suffix));
        
        if (isValidImage && !uploadedFiles.has(absPath)) {
          console.log(`[*] New file detected: ${filename}. Uploading...`);
          uploadedFiles.add(absPath); // optimistically add to prevent double upload
          
          try {
            // Wait 500ms to ensure file writing is complete
            await new Promise(r => setTimeout(r, 500));
            
            const fileBuffer = fs.readFileSync(filePath);
            const blob = new Blob([fileBuffer], { type: 'image/jpeg' });
            const formData = new FormData();
            formData.append('file', blob, filename);
            
            const response = await fetch(UPLOAD_URL, {
              method: 'POST',
              body: formData
            });
            
            const json = await response.json();
            
            if (response.ok) {
              console.log(`[✓] Successfully uploaded: ${filename}`);
              saveToHistory(absPath);
            } else {
              console.error(`[-] Upload failed for ${filename}. Status: ${response.status}`);
              console.error('    Response:', JSON.stringify(json));
              uploadedFiles.delete(absPath); // remove on failure to allow retry
            }
          } catch (uploadErr) {
            console.error(`[-] Error uploading ${filename}:`, uploadErr.message);
            uploadedFiles.delete(absPath); // remove on failure to allow retry
          }
        }
      }
    }
  } catch (err) {
    console.error('[-] Error in watch loop:', err.message);
  }
}, 1500);

File: upload_bridge.py
import os
import time
import requests
import mimetypes

# ==============================================================================
# CONFIGURATION
# ==============================================================================
# 1. Path to the folder where your camera uploads photos (via FTP, SD card, etc.)
#    Use double backslashes on Windows, e.g., r"C:\Camera\FTP" or "C:\\Camera\\FTP"
WATCH_FOLDER = r"./camera_photos"

# 2. Your API Upload URL (obtained from the dashboard integration panel)
#    Example: "http://localhost:3000/api/upload/your-event-id?secret=your-secret"
UPLOAD_URL = "http://localhost:3000/api/upload/replace-this-with-event-id?secret=your-secret"

# 3. File containing the log of already uploaded files (to prevent re-upload on restart)
HISTORY_FILE = ".uploaded_history.log"
# ==============================================================================

def load_history():
    if not os.path.exists(HISTORY_FILE):
        return set()
    try:
        with open(HISTORY_FILE, "r", encoding="utf-8") as f:
            return set(line.strip() for line in f if line.strip())
    except Exception as e:
        print(f"[-] Error loading history file: {e}")
        return set()

def save_to_history(filepath):
    try:
        with open(HISTORY_FILE, "a", encoding="utf-8") as f:
            f.write(filepath + "\n")
    except Exception as e:
        print(f"[-] Error writing to history file: {e}")

def main():
    print("=" * 60)
    print("           AURA STREAM - CAMERA UPLOAD BRIDGE (PYTHON)")
    print("=" * 60)
    
    # Ensure watch folder exists
    if not os.path.exists(WATCH_FOLDER):
        try:
            os.makedirs(WATCH_FOLDER)
            print(f"[+] Created watch folder: {os.path.abspath(WATCH_FOLDER)}")
        except Exception as e:
            print(f"[-] Error creating watch folder: {e}")
            return
    else:
        print(f"[+] Watching folder: {os.path.abspath(WATCH_FOLDER)}")
        
    print(f"[+] API Endpoint: {UPLOAD_URL}")
    print(f"[+] History stored in: {os.path.abspath(HISTORY_FILE)}")
    print("[*] Waiting for new photos... (Press Ctrl+C to stop)")
    print("-" * 60)

    uploaded_files = load_history()
    print(f"[*] Loaded {len(uploaded_files)} previously uploaded files from history.")

    while True:
        try:
            for filename in os.listdir(WATCH_FOLDER):
                # Ignore hidden files
                if filename.startswith('.'):
                    continue
                
                filepath = os.path.join(WATCH_FOLDER, filename)
                
                # Check if it is a file and a valid image
                if os.path.isfile(filepath):
                    ext = filename.lower()
                    if ext.endswith(('.jpg', '.jpeg', '.png', '.heic', '.heif', '.cr2', '.cr3', '.nef', '.arw')):
                        abs_path = os.path.abspath(filepath)
                        
                        if abs_path not in uploaded_files:
                            print(f"[*] New file detected: {filename}. Uploading...")
                            
                            # Simple retry mechanism
                            success = False
                            for attempt in range(3):
                                try:
                                    # Wait briefly for file to finish writing (especially for FTP uploads)
                                    time.sleep(0.5)
                                    
                                    mime_type, _ = mimetypes.guess_type(filepath)
                                    mime_type = mime_type or 'image/jpeg'
                                    
                                    with open(filepath, 'rb') as img:
                                        files = {'file': (filename, img, mime_type)}
                                        response = requests.post(UPLOAD_URL, files=files, timeout=30)
                                        
                                    if response.ok:
                                        print(f"[✓] Successfully uploaded: {filename}")
                                        uploaded_files.add(abs_path)
                                        save_to_history(abs_path)
                                        success = True
                                        break
                                    else:
                                        print(f"[-] Upload failed for {filename} (Attempt {attempt+1}/3). Status code: {response.status_code}")
                                        print(f"    Response: {response.text}")
                                except Exception as upload_err:
                                    print(f"[-] Error uploading {filename} (Attempt {attempt+1}/3): {upload_err}")
                                time.sleep(2)
                                
                            if not success:
                                print(f"[✗] Failed to upload {filename} after 3 attempts.")
                                
            time.sleep(1.5)
        except KeyboardInterrupt:
            print("\n[*] Bridge stopped. Goodbye!")
            break
        except Exception as e:
            print(f"[-] Error in watch loop: {e}")
            time.sleep(5)

if __name__ == "__main__":
    main()
